contentcuration/contentcuration/db/models/manager.py

```diff
@@ -41,21 +57,53 @@ def _get_next_tree_id(self, *args, **kwargs):
         new_id = MPTTTreeIDManager.objects.create().id
         return new_id
 
+    @contextlib.contextmanager
+    def _attempt_lock(self, tree_ids, values):
+        """
+        Internal method to allow the lock_mptt method to do retries in case of deadlocks
+        """
+        with transaction.atomic():
+            # Issue a separate lock on each tree_id
+            # in a predictable order.
+            # This will mean that every process acquires locks in the same order
+            # and should help to minimize deadlocks
+            for tree_id in tree_ids:
+                execute_queryset_without_results(
+                    self.select_for_update()
+                    .order_by()
+                    .filter(tree_id=tree_id)
+                    .values(*values)
+                )
+            yield
+
     @contextlib.contextmanager
     def lock_mptt(self, *tree_ids):
         # If this is not inside the context of a delay context manager
         # or updates are not disabled set a lock on the tree_ids.
         if not self.model._mptt_is_tracking and self.model._mptt_updates_enabled:
-            with transaction.atomic():
-                # Lock only MPTT columns for updates on any of the tree_ids specified
-                # until the end of this transaction
-                query = Q()
-                for tree_id in tree_ids:
-                    query |= Q(tree_id=tree_id)
-                self.select_for_update().order_by().filter(query).values(
-                    "tree_id", "lft", "rght"
-                )
-                yield
+            tree_ids = sorted((t for t in set(tree_ids) if t is not None))
+            # Lock based on MPTT columns for updates on any of the tree_ids specified
+            # until the end of this transaction
+            mptt_opts = self.model._mptt_meta
+            values = (
+                mptt_opts.tree_id_attr,
+                mptt_opts.left_attr,
+                mptt_opts.right_attr,
+                mptt_opts.level_attr,
+                mptt_opts.parent_attr,
+            )
+            try:
+                with self._attempt_lock(tree_ids, values):
+                    yield
+            except OperationalError as e:
+                if "deadlock detected" in e.args[0]:
+                    logging.error(
+                        "Deadlock detected while trying to lock ContentNode trees for mptt operations, retrying"
+                    )
+                    with self._attempt_lock(tree_ids, values):
```

Comment:
Wouldn't this call happen almost immediately after the call that raised the deadlock error? (I think PG waits 1 second to check for deadlock.) Have we found the conditions for deadlocks don't last longer than that, or is that the only scenario that we can reliably reproduce and handle with this code?

Reply:
Yes, it would happen immediately afterwards - and that's the point. By doing the call again immediately, there will be a lock due to the other process that triggered the deadlock, but no longer a deadlock with that process. The only possible danger here is that another process will trigger another deadlock, but I didn't want to do an infinite retry loop, so thought that would raise.

Reply:
I thought a deadlock would happen if a second process tried to obtain a lock on a row that has already been locked by another process? Am I wrong, and it only happens if two processes are trying to obtain the lock on the same row at the same time?

Reply:
A deadlock happens when two processes gain multiple locks in ways that conflict with each other. For example

Process A gains a lock on row 1.
Process B gains a lock on row 2.
Process B tries to gain a lock on row 1 - has to wait for A.

With its original lock on row 1 still engaged, Process A tries to gain a lock on row 2. It has to wait until Process B has finished.

Neither can now finish, because A is waiting for B to release its lock on 2, and B is waiting for A to release its lock on 1. They are deadlocked.

One of these processes will now abort due to the deadlock, and the other will continue. The one that aborts, we catch, and try again, knowing that the other processes has now acquire the locks it needs, so the chance of another deadlock with this process is much lower.

We do the pre-emptive locking for trees in a defined order in order to reduce this risk even further.

Reply:
Right, I guess what is tripping me up is that the query could trigger thousands (or even tens of thousands) of row locks, but I guess obtaining the lock for all of those rows must be quite fast such that even trying a moment later, the other process is very unlikely to still be locking rows. The reason I'm wondering about this is that since we are managing the locking ourselves, we have to IIUC lock the entire tree since we don't know which rows MPTT will affect (even when the number of actual affected rows could be quite small), so wanted to be sure that with larger trees that all that additional locking didn't increase the chances of deadlocks. 

In any case, it probably makes sense to merge this and just do more general testing with large trees with this in mind.